set.c

```diff
@@ -494,6 +494,14 @@ set_i_initialize(int argc, VALUE *argv, VALUE set)
     rb_check_arity(argc, 0, 1);
 
     if (argc > 0 && (other = argv[0]) != Qnil) {
+        ID id_size = rb_intern("size");
```

Comment:
Should move this to the following `else` block, so it won't be called for arrays, as arrays never have infinite size.